Move avo assembly generator to separate `asm` module

mmcloughlin opened the pull request:
Move `avo` code-generator to an `asm` sub-module. This avoids the problem of the `sha1cd` module itself depending on `avo`.

Fixes #148

pjbgf commented:
@mmcloughlin thank you for proposing this PR, that's much appreciated. Please amend the `make verify` target, so that it points to the new `asm/asm.go` which should fix the build.

mmcloughlin commented:
> @mmcloughlin thank you for proposing this PR, that's much appreciated. Please amend the `make verify` target, so that it points to the new `asm/asm.go` which should fix the build.

Thanks @pjbgf. I have pushed a new iteration with changes:

* Moved the `ubc` code generator to its own `asm` sub-module.
* Renamed the `ubc/doc.go` to `ubc/ubc.go` and added the `go:generate` line there too
* Updated `make generate` to just call `go generate -x ./...`
* Removed `sed` lines that add `// +build` constraints. I think `go:build` is sufficiently established that there's no need to keep the old form, unless you want to?

The only remaining thing I'd say is that the `shared` package doesn't make much sense anymore in this refactor at the moment. To avoid trying to share an internal package from the main module I just copied in the few constants needed by the assembly generator. So perhaps now `internal/const.go` should just be moved to the top-level package (and made private)?

mmcloughlin commented on asm/asm.go:
```diff
@@ -1,17 +1,22 @@
-//go:generate go run sha1cdblock_amd64_asm.go -out sha1cdblock_amd64.s
+const (
+	// Constants for the SHA-1 hash function.
+	RoundConst0 = 0x5A827999
+	RoundConst1 = 0x6ED9EBA1
+	RoundConst2 = 0x8F1BBCDC
+	RoundConst3 = 0xCA62C1D6
```
Note they can't be named `K0, K1, ...` anymore since these are registers in `avo` (AVX-512 mask registers).

pjbgf commented:
> So perhaps now internal/const.go should just be moved to the top-level package (and made private)?

That makes sense. Overall, the changes LGTM. It is up to you if you prefer making the additional change for the shared package here, or on a follow-up PR.

pjbgf reviewed (approved):
@mmcloughlin I will be merging this as is, thank you once again for proposing this change. The shared package change can be done at a later time.
